Libraries/pbxbuild/Sources/Phase/ProductTypeResolver.cpp

```diff
@@ -268,6 +269,135 @@ ResolveFrameworkStructure(Phase::Environment const &phaseEnvironment, Phase::Con
+    }
+
+    return ext::nullopt;
+}
+
+static bool
+ResolveModuleMap(Phase::Environment const &phaseEnvironment, Phase::Context *phaseContext, pbxsetting::Environment const &environment)
+{
+    std::string const &workingDirectory = phaseContext->toolContext().workingDirectory();
+    // TODO: what about non-framework product types?
+    std::string finalModuleMapRoot = environment.resolve("TARGET_BUILD_DIR") + "/" + environment.resolve("CONTENTS_FOLDER_PATH") + "/" + "Modules";
+
+    /*
+     * Handle the standard module map, either copied or generated.
+     */
+    std::string inputModuleMapPath = environment.resolve("MODULEMAP_FILE");
+    std::string intermediateModuleMapPath = environment.resolve("TARGET_TEMP_DIR") + "/" + "module.modulemap";
+    ext::optional<Tool::Invocation::AuxiliaryFile> moduleMapAuxiliaryFile;
+    if (!inputModuleMapPath.empty()) {
+        /*
+         * Copy in the input module map as an auxiliary file.
+         */
+        moduleMapAuxiliaryFile = Tool::Invocation::AuxiliaryFile(
+            intermediateModuleMapPath,
+            FSUtil::ResolveRelativePath(inputModuleMapPath, workingDirectory),
+            false);
+    } else {
+        /*
+         * Find umbrella header: a header with the same name as the target's module.
+         */
+        std::string moduleName = environment.resolve("PRODUCT_MODULE_NAME");
+        ext::optional<std::string> umbrellaHeaderName = FindUmbrellaHeaderName(environment, moduleName, phaseEnvironment.target());
+
+        if (umbrellaHeaderName) {
+            /*
+             * Generate a module map as an auxiliary file.
+             */
+            auto contents = std::string();
+            contents += "framework module " + moduleName + " {\n";
+            contents += "  umbrella header \"" + *umbrellaHeaderName + "\"\n";
+            contents += "  \n";
+            contents += "  export *\n";
+            contents += "  module * { export * }\n";
+            contents += "}\n";
+
+            moduleMapAuxiliaryFile = Tool::Invocation::AuxiliaryFile(
+                intermediateModuleMapPath,
+                std::vector<uint8_t>(contents.begin(), contents.end()),
+                false);
+        } else {
+            fprintf(stderr, "warning: target defines module, but has no umbrella header\n");
+        }
+    }
+
+    if (moduleMapAuxiliaryFile) {
+        /* Define source module map. */
+        // TODO: it would be nicer to attach this to the copy invocation created below
+        Tool::Invocation invocation;
+        invocation.auxiliaryFiles().push_back(*moduleMapAuxiliaryFile);
+        phaseContext->toolContext().invocations().push_back(invocation);
+
+        /* Copy in the module map as part of the build. */
+        if (Tool::CopyResolver const *copyResolver = phaseContext->copyResolver(phaseEnvironment)) {
+            copyResolver->resolve(&phaseContext->toolContext(), environment, { intermediateModuleMapPath }, finalModuleMapRoot, "Ditto");
```

Comment:
The example log I was basing this on said "Ditto" in the log message. Not sure why, since it's not actually using the `/usr/bin/ditto` tool, but I guess since it's just a simple copy.